File: Source/Tests/CSharp/Extensions/FieldSymbolExtensions.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Tests.CSharp.Extensions
{
	using System;

	namespace FieldSymbolExtensionsTests
	{
		using System.Linq;
		using FluentAssertions;
		using Microsoft.CodeAnalysis;
		using Microsoft.CodeAnalysis.CSharp;
		using Microsoft.CodeAnalysis.CSharp.Syntax;
		using NUnit.Framework;
		using SafetySharp.CSharp.Extensions;

		[TestFixture]
		internal class GetFullNameMethod
		{
			private static string GetFullName(string csharpCode)
			{
				var compilation = new TestCompilation(csharpCode);
				var typeDeclaration = compilation
					.SyntaxRoot
					.DescendantNodesAndSelf<TypeDeclarationSyntax>()
					.Single(c => c.Identifier.ValueText == "X");

				var typeSymbol = compilation.SemanticModel.GetDeclaredSymbol(typeDeclaration);
				return typeSymbol.GetMembers().OfType<IFieldSymbol>().Single(field => field.Name == "f").GetFullName();
			}

			[Test]
			public void ReturnsFieldNameForGenericType()
			{
				GetFullName("class X<T1, T2> { int f; }").Should().Be("X<T1, T2>.f");
				GetFullName("struct X<T1, T2> { int f; }").Should().Be("X<T1, T2>.f");
			}

			[Test]
			public void ReturnsFieldNameForNestedTypes()
			{
				GetFullName("class Y { class X { int f; }}").Should().Be("Y+X.f");
				GetFullName("namespace Test.Other { class Y { class X { int f; }} }").Should().Be("Test.Other.Y+X.f");
				GetFullName("namespace Test.Other { struct Y { class X { int f; }} }").Should().Be("Test.Other.Y+X.f");
				GetFullName("namespace Test.Other { class Y { struct X { int f; }} }").Should().Be("Test.Other.Y+X.f");
				GetFullName("namespace Test.Other { struct Y { struct X { int f; }} }").Should().Be("Test.Other.Y+X.f");
				GetFullName("namespace Test { namespace Other { class Y { class X { int f; }} }}").Should().Be("Test.Other.Y+X.f");
				GetFullName("namespace Test { class Y { class X { int f; }} }").Should().Be("Test.Y+X.f");
			}

			[Test]
			public void ReturnsFieldNameForTypeInGlobalNamespace()
			{
				GetFullName("class X { int f; }").Should().Be("X.f");
				GetFullName("struct X { int f; }").Should().Be("X.f");
			}

			[Test]
			public void ReturnsFieldNameForTypeInNamespace()
			{
				GetFullName("namespace Test { class X { private int f; } }").Should().Be("Test.X.f");
				GetFullName("namespace Test.Other { class X { int f; } }").Should().Be("Test.Other.X.f");
				GetFullName("namespace Test { namespace Other { class X { int f; } }}").Should().Be("Test.Other.X.f");
				GetFullName("namespace Test { struct X { private int f; } }").Should().Be("Test.X.f");
				GetFullName("namespace Test.Other { struct X { int f; } }").Should().Be("Test.Other.X.f");
				GetFullName("namespace Test { namespace Other { struct X { int f; } }}").Should().Be("Test.Other.X.f");
			}
		}
	}
}

File: Tests/CSharp/Roslyn/Symbols/FieldSymbolExtensions.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Roslyn.Symbols.FieldSymbolExtensions

open System
open System.Linq
open NUnit.Framework
open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax
open SafetySharp.CSharpCompiler.Roslyn.Symbols

[<TestFixture>]
module ``IsSubcomponentField methods`` =
    let isComponentField csharpCode =
        let compilation = TestCompilation csharpCode

        let classSymbol = compilation.CSharpCompilation.GetTypeSymbol "X"
        let fieldSymbol = classSymbol.GetMembers().OfType<IFieldSymbol>().Single()
            
        fieldSymbol.IsSubcomponentField compilation.SemanticModel && fieldSymbol.IsSubcomponentField compilation.CSharpCompilation

    [<Test>]
    let ``throws if field symbol is null`` () =
        let compilation = TestCompilation ""
        raisesArgumentNullException "fieldSymbol" (fun () -> (null : IFieldSymbol).IsSubcomponentField compilation.CSharpCompilation |> ignore)
        raisesArgumentNullException "fieldSymbol" (fun () -> (null : IFieldSymbol).IsSubcomponentField compilation.SemanticModel |> ignore)

    [<Test>]
    let ``throws if semantic model is null`` () =
        let compilation = TestCompilation "class C { int x; }"
        let fieldSymbol = compilation.FindFieldSymbol "C" "x"
        raisesArgumentNullException "semanticModel" (fun () -> fieldSymbol.IsSubcomponentField (null : SemanticModel) |> ignore)

    [<Test>]
    let ``throws if compilation is null`` () =
        let compilation = TestCompilation "class C { int x; }"
        let fieldSymbol = compilation.FindFieldSymbol "C" "x"
        raisesArgumentNullException "compilation" (fun () -> fieldSymbol.IsSubcomponentField (null : Compilation) |> ignore)

    [<Test>]
    let ``returns false for non-component fields`` () =
        isComponentField "class X : Component { int x; }" =? false
        isComponentField "class X : Component { bool x; }" =? false
        isComponentField "class X : Component { decimal x; }" =? false

    [<Test>]
    let ``returns true for component fields`` () =
        isComponentField "class X : Component { Component x; }" =? true
        isComponentField "class X : Component { IComponent x; }" =? true
        isComponentField "class Y : Component {} class X : Component { Y x; }" =? true
        isComponentField "interface Y : IComponent {} class X : Component { Y x; }" =? true